parser/src/token.rs

```diff
@@ -1,108 +1,669 @@
+        let location = self.eof_location;
+        self.error(location, UnexpectedEof)
+    }
+
+    fn slice(&self, start: Location, end: Location) -> &'input str {
+        let start = start.absolute;
+        let end = end.absolute;
+
+        &self.input[start.to_usize()..end.to_usize()]
+    }
+
+    fn take_while<F>(&mut self, start: Location, mut keep_going: F) -> (Location, &'input str)
+        where F: FnMut(char) -> bool,
+    {
+        self.take_until(start, |c| !keep_going(c))
+    }
+
+    fn take_until<F>(&mut self, start: Location, mut terminate: F) -> (Location, &'input str)
+        where F: FnMut(char) -> bool,
+    {
+        while let Some((end, ch)) = self.lookahead {
+            if terminate(ch) {
+                return (end, self.slice(start, end));
+            } else {
+                self.bump();
+            }
+        }
+        (self.eof_location, self.slice(start, self.eof_location))
+    }
+
+    fn test_lookahead<F>(&self, mut test: F) -> bool
+        where F: FnMut(char) -> bool,
+    {
+        self.lookahead.map_or(false, |(_, ch)| test(ch))
+    }
+
+    fn line_comment(&mut self, start: Location) -> Option<SpannedToken<'input>> {
+        let (end, comment) = self.take_until(start, |ch| ch == '\n');
+
+        if comment.starts_with("///") {
+            let skip = if comment.starts_with("/// ") { 4 } else { 3 };
+            let doc = Token::DocComment(comment[skip..].to_string());
+            Some(pos::spanned2(start, end, doc))
+        } else {
+            None
+        }
+    }
+
+    fn block_comment(&mut self, start: Location) -> Result<Option<SpannedToken<'input>>, Error> {
+        self.bump(); // Skip first '*'
+
+        loop {
+            let (_, comment) = self.take_until(start, |ch| ch == '*');
+            self.bump(); // Skip next '*'
+            match self.lookahead {
+                Some((end, '/')) => {
+                    self.bump();
+                    if comment.starts_with("/**") && comment != "/**" {
+                        // FIXME: whitespace alignment
+                        let doc = Token::DocComment(comment[3..].trim_left().to_string());
+                        return Ok(Some(pos::spanned2(start, end.shift('/'), doc)));
+                    } else {
+                        return Ok(None);
+                    }
+                }
+                Some((_, _)) => continue,
+                None => return self.eof_error(),
+            }
+        }
+    }
+
+    fn operator(&mut self, start: Location) -> SpannedToken<'input> {
+        let (end, op) = self.take_while(start, is_operator_char);
+
+        let token = match op {
+            "." => Token::Dot,
+            ":" => Token::Colon,
+            "=" => Token::Equals,
+            "|" => Token::Pipe,
+            "->" => Token::RArrow,
+            "#" => {
+                // Is this too permissive?
+                self.take_while(start, is_ident_start);
```

Comment:
I believe this is the same logic as before. It is not ideal but it works #204 or something similiar to make this more solid would be nice however.